packages/iov-core/src/jsrpc.ts

```diff
@@ -0,0 +1,242 @@
+import {
+  jsonRpcCodeInternalError,
+  jsonRpcCodeInvalidParams,
+  jsonRpcCodeInvalidRequest,
+  jsonRpcCodeMethodNotFound,
+  jsonRpcCodeParseError,
+  jsonRpcCodeServerErrorDefault,
+  SimpleMessagingConnection,
+} from "@iov/jsonrpc";
+import { firstEvent } from "@iov/stream";
+
+// An RPC interface very similar to JSON-RPC but supporting
+// all the data types from the structured clone algorithm
+// https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm#Supported_types
+
+/** A single structured clone algorithm compatible value */
+export declare type JsRpcCompatibleValue =
+  | string
+  | number
+  | boolean
+  | null
+  | Uint8Array
+  | JsRpcCompatibleArray
+  | JsRpcCompatibleDictionary;
+
+/** An array of JsRpcCompatibleValue */
+export interface JsRpcCompatibleArray extends ReadonlyArray<JsRpcCompatibleValue> {}
+
+/** A string to JsRpcCompatibleValue dictionary. */
+export interface JsRpcCompatibleDictionary {
+  readonly [key: string]: JsRpcCompatibleValue;
+}
+
+export function isJsRpcCompatibleValue(value: unknown): value is JsRpcCompatibleValue {
+  return (
+    typeof value === "string" ||
+    typeof value === "number" ||
+    typeof value === "boolean" ||
+    value === null ||
+    value instanceof Uint8Array ||
+    isJsRpcCompatibleArray(value) ||
+    isJsRpcCompatibleDictionary(value)
+  );
+}
+
+export function isJsRpcCompatibleArray(value: unknown): value is JsRpcCompatibleArray {
+  if (!Array.isArray(value)) {
+    return false;
+  }
+  return value.every(isJsRpcCompatibleValue);
+}
+
+export function isJsRpcCompatibleDictionary(data: unknown): data is JsRpcCompatibleDictionary {
+  if (typeof data !== "object" || data === null) {
+    // data must be a non-null object
+    return false;
+  }
+
+  // Exclude special kind of objects like Array, Date or Uint8Array
+  // Object.prototype.toString() returns a specified value:
+  // http://www.ecma-international.org/ecma-262/7.0/index.html#sec-object.prototype.tostring
+  if (Object.prototype.toString.call(data) !== "[object Object]") {
+    return false;
+  }
+
+  for (const key of Object.getOwnPropertyNames(data)) {
+    const value = (data as any)[key];
+    if (!isJsRpcCompatibleValue(value)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+export interface JsRpcRequest {
+  readonly id: number;
+  readonly method: string;
+  readonly params: JsRpcCompatibleArray | JsRpcCompatibleDictionary;
```

Comment:
:smile: supporting both formats from the spec

Reply:
Correct. We only use `JsRpcCompatibleDictionary` in our RPC interface but due to separation of concerns, the type allows both.